<HTML><HEAD><META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
<!--SafClassName="docSection1Title"--><!--SafTocEntry="A.2 Built-in Exceptions"-->
<LINK REL="stylesheet" href="FILES/sbolM.css">
</HEAD>
<BODY link="#354278" vlink="#000066" alink="#000080" topmargin="0">
<table width="100%" border="0" cellspacing="0" cellpadding="0" bgcolor="#CCCCCC">
<td><font color="black" size=1>I l<font color="#FF0000">@</font>ve RuBoard</td>
<td valign="top" class="v2" align="right"><a href="pythonwin32_snode182.html"><img src="FILES/previous.gif" width="62" height="15" border="0" align="absmiddle" alt="Previous Section"></a> <a href="pythonwin32_snode184.html"><img src="FILES/next.gif" width="41" height="15" border="0" align="absmiddle" alt="Next Section"></a></td></table>
<FONT>
<h3>A.2
Built-in Exceptions </h3>


<p>
<a NAME="idx2181"></a>Exceptions can be
<tt cLASS="monofont">class</tt> or <tt CLASs="monofont">string</tt> objects. While
most exceptions have traditionally been <tt class="monofont">string</tt>
objects, in Python 1.5 all standard exceptions have been converted to
class objects, and users are encouraged to do the same. The source
code for those exceptions is present in the standard library-module
exceptions; this module never needs to be imported explicitly.</p>



<p>For backward compatibility, when Python is invoked with the
-<tt class="monofont">X</TT> option, most standard exceptions are
strings.<FONT sIZe="1"><SUp><A HREF="#FOOTNOTE-F">[F]</A></sup></fONT>
 This
option can run code that breaks because of the different semantics of
class-based exceptions. The -<Tt claSS="monofont">X</TT> option will become
obsolete in future Python versions, so the recommended solution is to
fix the code.</p>


<bloCKQUote><font size="1">
<p class="footnote">
<sup><a NAME="FOOTNOTE-F">[F]</A></SuP>
For forward-compatibility the new exceptions
<Tt CLaSS="monofont">Exception</TT>, <TT clasS="monofont">LookupError</TT>,
<Tt claSS="monofont">ArithmeticError</TT>,
<tt clASS="monofont">EnvironmentError</Tt>, and
<tt class="monofont">StandardError</tt> are tuples.</p>
</font></blockqUOTE>





<P>Two distinct string objects with the same value are considered
different exceptions. This forces programmers to use exception names
rather than their string value when specifying exception handlers.
The string value of all built-in exceptions is their name, but this
isn't a requirement for user-defined exceptions or exceptions
defined by library modules.</P>



<p>For class exceptions, in a <TT cLAsS="monofont">try</TT> statement with an
except clause that mentions a particular class, that clause also
handles any exception classes derived from that class (but not
exception classes from which it is derived). Two exception classes
that aren't related via subclassing are never equivalent, even
if they have the same name.</P>



<P>The built-in exceptions in the following list can be generated by the
interpreter or built-in functions. Except where mentioned, they have
an "associated value" indicating the detailed cause of
the error. This may be a string or a tuple containing several items
of information (e.g., an error code and a string explaining the
code). The associated value is the second argument to the raise
statement. For string exceptions, the associated value itself is
stored in the variable named as the second argument of the
<Tt claSS="monofont">except</TT> clause (if any). For class exceptions, that
variable receives the exception instance. If the exception class is
derived from the standard root class <tt clASS="monofont">Exception</Tt>,
the associated value is present as the exception instance's
<tt cLASS="monofont">args</tt> attribute, and possibly on other attributes
as well.</p>



<p>User code can raise built-in exceptions. This code can test an
exception handler or report an error condition "just
like" the situation in which the interpreter raises the same
exception; but beware that there is nothing to prevent user code from
raising an inappropriate error.</p>



<p>The following exceptions are used only as base classes for other
exceptions. When string-based standard exceptions are used, they are
tuples containing the directly derived classes:</p>



<dl>
<dt>
<font color="#990000"><fONT COLoR="#990000"><SpAN cLASS="monofont">Exception</SPan></foNT>
 </FOnt></dt>
<DD>
<P>The root class for exceptions. All built-in exceptions are derived
from this class. All user-defined exceptions should also be derived
from this class, but this isn't (yet) enforced. The
<Tt claSS="monofont">str()</TT> function, when applied to an instance of
this class (or most derived classes) returns the string value of the
argument or arguments, or an empty string if no arguments were given
to the constructor. When used as a sequence, this accesses the
arguments given to the constructor (handy for backward compatibility
with old code). The arguments are also available on the
instance's <tt class="monofont">args</tt> attribute, as a tuple.</p>
</dd>
<dt>
<font colOR="#990000"><FONT cOLoR="#990000"><SpAN CLASs="monofont">StandardError</spaN></FONt>
 </fonT></DT>
<Dd>
<p>The base class for all built-in exceptions except
<tt CLASs="monofont">SystemExit</tt>. <tt class="monofont">StandardError</tt>
itself is derived from the root class <tt class="monofont">Exception</tt>.</P>
</DD>
<DT>
<FoNT cOLoR="#990000"><FONT Color="#990000"><SPAN clasS="monofont">ArithmeticError</SPAn></fonT>
 </FONt></dt>
<dd>
<p>The base class for those built-in exceptions that are raised for
various arithmetic errors: <tt class="monofont">OverflowError</tt>,
<tt clasS="monofont">ZeroDivisionError</TT>,
<TT ClASs="monofont">FloatingPointError</TT>.</p>
</DD>
<DT>
<FOnt coLOR="#990000"><Font cOLOR="#990000"><span CLASs="monofont">LookupError</span></font>
 </font></dt>
<dd>
<p>The base class for the exceptions that are raised when a key or index
used on a mapping or sequence is invalid:
<tt cLASS="monofont">IndexError</TT>, <tT ClASs="monofont">KeyError</TT>.</P>
</DD>
<Dt>
<fonT COLor="#990000"><foNT COlor="#990000"><sPAN Class="monofont">EnvironmentError</span></font>
 </font></dt>
<dd>
<p>The base class for exceptions that can occur outside the Python
system: <TT CLASs="monofont">IOError</TT>, <tT ClASS="monofont">OSError</TT>. When
exceptions of this type are created with a two-tuple, the first item
is available on the instance's <Tt claSS="monofont">errno</TT>
attribute (it's assumed to be an error number), and the second
item is available on the <tt clASS="monofont">strerror</Tt> attribute
(it's usually the associated error message). The tuple itself
is also available on the <tt cLASS="monofont">args</tt> attribute. New in
Version 1.5.2.</p>



<p>When an <tt class="monofont">EnvironmentError</tt> exception is instantiated
with a three-tuple, the first two items are available as above, while
the third item is available on the filename attribute. However, for
backward-compatibility, the <tt class="monofont">args</tT> attribute
contains only a two-tuple of the first two constructor arguments.</P>



<P>The filename attribute is <TT ClASs="monofont">None</TT> when this exception
is created with other than three arguments. The
<tT CLASS="monofont">errno</tt> and <tt CLASs="monofont">strerror</tt> attributes
are also <tT CLAss="monofont">None</tt> if the instance was created with
other than two or three arguments. In this last case,
<TT CLass="monofont">args</tt> contains the verbatim constructor arguments
as a tuple.</p>
</dd>
</dl>


<p>The following exceptions are those actually raised. They are class
objects, except when the <tt class="monofont">-X</tt> option is used to
revert back to string-based standard exceptions:</p>



<A NAME="idx2182"></A><dL>
<Dt>
<FOnT COLOR="#990000"><font COLOr="#990000"><spaN CLAss="monofont">AssertionError</spAN></FOnt>
</font></dt>
<dd>
<p>Raised when an assert statement fails.</p>
</dd>
<dt>
<font cOLOR="#990000"><FOnT CoLOr="#990000"><SPAN CLass="monofont">AttributeError</sPAN></Font>
 </fONT></Dt>
<dd>
<p>Raised when an attribute reference or assignment fails. (When an
object doesn't support attribute references or attribute
assignments at all, <TT CLass="monofont">TypeError</tt> is raised.)</p>
</dd>
<dt>
<font color="#990000"><foNT COLOr="#990000"><SPaN ClASS="monofont">EOFError</SPAn></fonT>
 </FONt></dt>
<dD>
<P>Raised when one of the built-in functions (<TT clasS="monofont">input()</TT>
or <Tt class="monofont">raw_input()</tt>) hits an end-of-file condition
(EOF) without reading any data. (Note that the
<tt class="monofont">read()</tt> and <tt clASS="monofont">readline()</TT> methods
of file objects return an empty string when they hit EOF.)</P>
</dD>
<Dt>
<FOnT COLOR="#990000"><font COLOr="#990000"><spaN CLAss="monofont">FloatingPointError</spAN></FOnt>
 </font></dt>
<dd>
<p>Raised when a floating-point operation fails. This exception is
always defined, but can be raised only when Python is configured with
the <tt class="monofont">-with-fpectl</tt> option or the
<tT CLASS="monofont">WANT_SIGFPE_HANDLER</tT> symbol is defined in the
<I>config.h</i> file.</P>
</Dd>
<DT>
<FONT coloR="#990000"><FONt colOR="#990000"><SPan clASS="monofont">IOError</Span></font>
 </font></dt>
<dd>
<p>Raised when an I/O operation (such as a print statement, the built-in
<tt claSS="monofont">open()</TT> function, or a method of a file object)
fails for an I/O-related reason, e.g., file not found or disk full.</P>



<P>This class is derived from <tT ClASs="monofont">EnvironmentError</TT>. See
its previous discussion for more information on exception-instance
attributes.</P>
</DD>
<Dt>
<fonT COLor="#990000"><foNT COlor="#990000"><sPAN Class="monofont">ImportError</span></font>
 </font></dt>
<dd>
<p>Raised when an import statement fails to find the module definition
or when a <TT CLASs="monofont">from</TT> <tT ClASS="monofont">...</TT>
<Tt claSS="monofont">import</TT> fails to find a name that's to be
imported.</p>
</dd>
<dT>
<FONt colOR="#990000"><FOnt color="#990000"><span class="monofont">IndexError</span></fONT>
 </FONt></DT>
<dD>
<P>Raised when a sequence subscript is out of range. (Slice indexes are
silently truncated to fall in the allowed range; if an index
isn't a plain integer, <tT CLASS="monofont">TypeError</tt> is raised.)</p>
</dD>
<DT>
<Font cOLOR="#990000"><font COLOr="#990000"><span class="monofont">KeyError</span></font>
 </fonT></DT>
<DD>
<P>Raised when a mapping (dictionary) key is not found in the set of
existing keys.</p>
</DD>
<dT>
<FoNT COLOr="#990000"><fonT COLor="#990000"><spAN CLass="monofont">KeyboardInterrupt</sPAN></Font>
 </font></dt>
<dd>
<p>Raised when the user hits the interrupt key (normally Ctrl-C or Del).
During execution, a check for interrupts is made regularly.
Interrupts typed when a built-in function <tt class="monofont">input()</tt>
or <TT CLASs="monofont">raw_input()</TT>) is waiting for input also raise
this exception.</p>
</DD>
<dT>
<FONT Color="#990000"><FONT coloR="#990000"><SPAn claSS="monofont">MemoryError</SPan></font>
 </font></dt>
<dd>
<p>Raised when an operation runs out of memory but the situation may
still be rescued (by deleting some objects). The associated value is
a string indicating what kind of (internal) operation ran out of
memory. Note that because of the underlying memory-management
architecture (C's <tt clasS="monofont">malloc()</TT> function), the
interpreter may not always completely recover from this situation; it
nevertheless raises an exception so that a stack traceback can be
printed, in case a runaway program was the cause.</P>
</DD>
<dT>
<FoNT cOLOR="#990000"><FOnt coLOR="#990000"><Span cLASS="monofont">NameError</span></FONT>
 </font></dt>
<dd>
<p>Raised when a local or global name is not found. Applies only to
unqualified names. The associated value is the name that can't
be found.</p>
</dd>
<dt>
<font colOR="#990000"><FONT cOLoR="#990000"><SpAN CLASs="monofont">NotImplementedError</spaN></FONt>
 </fonT></DT>
<Dd>
<p>Derived from <tt CLASs="monofont">RuntimeError</tt>. In user-defined base
classes, abstract methods should raise this exception when they
require derived classes to override the method. New in Version 1.5.2.</p>
</dd>
<dt>
<font color="#990000"><font COLOR="#990000"><SpAN cLAsS="monofont">OSError</SPAN></Font>
 </fONT></Dt>
<dd>
<p>Derived from <TT CLass="monofont">EnvironmentError</tT> and is used
primarily as the <TT Class="monofont">os</tt> module's
<tt class="monofont">os.error</tt> exception. See
<tt clasS="monofont">EnvironmentError</TT> in the first exception list for a
description of the possible associated values. New in Version 1.5.2.</P>
</DD>
<dT>
<FoNT cOLOR="#990000"><FOnt coLOR="#990000"><Span cLASS="monofont">OverflowError</span></FONT>
 </font></dt>
<dd>
<p>Raised when the result of an arithmetic operation is too large to be
represented. This can't occur for long integers (which would
rather raise <tt class="monofont">MemoryError</tt> than give up). Because of
the lack of standardization of floating-point exception handling in
C, most floating-point operations also aren't checked. For
plain integers, all operations that can overflow are checked except
left shift, where typical applications prefer to drop bits than raise
an exception.</p>
</dd>
<DT>
<FONT cOLoR="#990000"><FoNT COLOr="#990000"><spaN CLAss="monofont">RuntimeError</spAN></FOnt>
 </foNT></DT>
<dd>
<p>Raised when an error is detected that doesn't fall in any of
the other categories. The associated value is a string indicating
what precisely went wrong. (This exception is mostly a relic from a
previous version of the interpreter; it isn't used much any
more.)</p>
</dd>
<dt>
<font color="#990000"><font COLOR="#990000"><SpAN cLAsS="monofont">SyntaxError</SPAN></Font>
 </fONT></Dt>
<dd>
<p>Raised when the parser encounters a syntax error. This may occur in
an <TT CLass="monofont">import</tT> statement, in an <TT Class="monofont">exec</tt>
statement, in a call to the built-in function
<tt class="monofont">eval()</tt> or <tt clasS="monofont">input()</TT>, or when
reading the initial script or standard input (also interactively).</P>



<P>When class exceptions are used, instances of this class have
attributes <Tt CLaSS="monofont">filename</tT>, <TT CLAss="monofont">lineno</tt>,
<TT CLass="monofont">offset</tT>, and <TT Class="monofont">text</TT> for easier
access to the details; for string exceptions, the associated value is
usually a tuple of the form <TT class="monofont"><i>(message,</i></tt>
<tt class="monofont"><i>(filename,</i></tt>
<tT CLASS="monofont"><i>lineno,</I></Tt> <TT cLASS="monofont"><I>offset,</I></tt>
<tt CLASs="monofont"><i>text)</i></tT>. For class exceptions,
<TT Class="monofont">str()</TT> returns only the message.</P>
</Dd>
<dt>
<font color="#990000"><font color="#990000"><SPAN CLaSS="monofont">SystemError</sPAn></FONT>
 </FOnt></dt>
<DD>
<P>Raised when the interpreter finds an internal error, but the
situation doesn't look so serious to cause it to abandon all
hope. The associated value is a string indicating what went wrong (in
low-level terms). You should report this to the author or maintainer
of your Python interpreter. Be sure to report the version string of
the Python interpreter (<Tt claSS="monofont">sys.version</TT>, also printed
at the start of an interactive Python session), the exact error
message (the exception's associated value), and, if possible,
the source of the program that triggered the error.</p>
</dd>
<dT>
<FONt color="#990000"><font color="#990000"><span clASS="monofont">SystemExit</SPAn></FOnT>
 </FoNT></DT>
<DD>
<p>Raised by the <tt cLASS="monofont">sys.exit()</tt> function. When it's
not handled, the Python interpreter exits; no stack traceback is
printed. If the associated value is a plain integer, it specifies the
system exit status (passed to C's <tt CLASs="monofont">exit()</tt>
function); if it's <tT CLAss="monofont">None</tt>, the exit status is
zero; if it has another type (such as a string), the object's
value is printed, and the exit status is one.</p>



<p>When class exceptions are used, the instance has an attribute code
that is set to the proposed exit status or error message (defaulting
to <tt class="monofont">None</tt>). Also, this exception derives directly
from <tt clasS="monofont">Exception</TT> and not
<TT ClASs="monofont">StandardError</TT>, since it isn't technically an
error. A call to <tT CLASS="monofont">sys.exit()</tt> is translated into an
exception so that clean-up handlers (finally clauses of
<tt CLASs="monofont">try</tt> statements) can be executed, and so that a
debugger can execute a script without running the risk of losing
control. The <tT CLAss="monofont">os._exit()</tt> function can be used if
it's absolutely necessary to exit immediately (e.g., after a
<TT CLass="monofont">fork()</tt> in the child process).</p>
</dd>
<dt>
<font color="#990000"><foNT COLOr="#990000"><SPaN ClASS="monofont">TypeError</SPAn></fonT>
 </FONt></dt>
<dD>
<P>Raised when a built-in operation or function is applied to an object
of inappropriate type. The associated value is a string giving
details about the type mismatch.</P>
</Dd>
<dt>
<fONT Color="#990000"><font color="#990000"><span clasS="monofont">ValueError</SPAN></FoNT>
 </fONt></DT>
<DD>
<P>Raised when a built-in operation or function receives an argument
that has the right type but an inappropriate value, and the situation
is not described by a more precise exception such as
<Tt claSS="monofont">IndexError</TT>.</p>
</dd>
<dT>
<FONt colOR="#990000"><FOnt color="#990000"><span class="monofont">ZeroDivisionError</span></fONT>
 </FONt></DT>
<dD>
<P>Raised when the second argument of a division or modulo operation is
zero. The associated value is a string indicating the type of the
operands and the operation.<a NAME="idx2182"></A></P>
</dd>
</dl>
</FONT>
 
<table width="100%" border="0" cellspacing="0" cellpadding="0" bgcolor="#CCCCCC">
<td><font color="black" size=1>I l<font color="#FF0000">@</font>ve RuBoard</td>
<td valign="top" class="v2" align="right"><a href="pythonwin32_snode182.html"><img src="FILES/previous.gif" width="62" height="15" border="0" align="absmiddle" alt="Previous Section"></a> <a href="pythonwin32_snode184.html"><img src="FILES/next.gif" width="41" height="15" border="0" align="absmiddle" alt="Next Section"></a></td></table>
</BODY></HTML>